### ISSUE TITLE
expiring value analysis and begin blocks

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#14750]
Issue #13704 describes rules for identifying variables that can be de-initialized before the end of the block. Implementation of these rules to change the deinit point in PR #14691 revealed that the analysis did not consider `begin` statements but it arguably should.

For example:

``` chapel
proc test() {
  var r: R;
  var step1:sync int;
  var step2:sync int;

  begin {
    step1; // wait for 1st task
    writeln("task one ", r);
    step2 = 1;
  }

  begin {
    writeln("task two ", r);
    step1 = 1;
  }

  step2;
}
```

The compiler might not be aware that the tasks run in the reverse order that they are written in. We wouldn't want the record `r` to be destroyed within task two before task one runs, for example.

PR #14691 takes the approach of making variables that are referred to in `begin` blocks always be considered to expire at the end of the function. But this is arguably too conservative because `begin` blocks within a `sync` block would not be a problem. For example:

``` chapel
proc testSync() {
  var x: SomeRecord = new SomeRecord();
  sync {
    begin {
      f(x);
      // point 1
    }
  }
  // point 2
  g();
  // point 3
}
```

Here we almost certainly do not want do deinit x at point 1 (for reasons described above). We could however choose to deinit it at point 2 (just after the sync block) or point 3 in this particular case.

So, the options are:
 1. Any use of `begin` causes captured variables to be deinited at end of block
 2. `begin` within a `sync` will cause variables to be "last mentioned" just after the sync block
 3. Do not consider variables passed to `begin` to have impact on lifetime and instead make programs like `test` above some sort of error (either user error or compilation error)

**Comment by bradcray** [comment]
I like approach 2 when it applies, falling back to either case 1 or 3 when it doesn't (i.e., when the begins don't have a sync block at this lexical scope).  I can't decide whether 1 or 3 is more appropriate (e.g., should the lifetime checker be wrestling with cases like this, in which case maybe it takes care of case 3; or should it be on the user's shoulders, in which case case 1 seems OK).

**Comment by mppf** [comment]
I'm not so sure that approach 3 is reasonable - it would effectively be disallowing the `test` program shown above - meaning that `begin` with explicit synchronization wouldn't work anymore. That seems counter to the language design idea of having "fire and forget" begins in the first place.

When working on PR #14691, I ran into the problem that the compiler was adding the deinits before the begin task even started, e.g.:

``` chapel
proc testBegin() {
  var r: R = new R();
  var waitHere:sync int;
  begin {
     f(x)
     waitHere = 1;
  }
  // x was being deinited here, before the task starts
  waitHere;
  // at least we would need x deinited here to support such a pattern
}
```

except that instead of a sync variable it was the compiler's task counting (for sync blocks, etc).
But in any case the `testBegin` example above seems like something that should be a legal program... I suppose we could argue that the user has to explicitly extend the lifetime of `x` by adding a mention after `waitHere`, but... that seems to be adding unnecessary traps to the language IMO.

Note that we do currently have a TODO to extend lifetime checking to check that a begin does not refer to variables after they are no longer in scope. So we could hope for a compiler error in the above programs, but I still think that'd be making the language harder than it needs to be.

**Comment by vasslitvinov** [comment]
> I'm leaning towards approach 2 and falling back to case 1.

This is my preference as well.

**Comment by mppf** [comment]
Addressed by PR #14795.
